### PR TITLE
[REEF-1454] onResourceLaunch should not throw exception if evaluator failed

### PR DESCRIPTION
**Pull request by tcNickolas**
Evaluator can fail before task is submitted; there will be a FailedEvaluator
event produced which can be handled properly, and no need to throw exception
to crash process.

JIRA:
  [REEF-1454](https://issues.apache.org/jira/browse/REEF-1454)

Pull request:
  This closes #

### CODE REVIEW

**Comment by jwang98052** [comment]
Looks good to me.

File: lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/driver/evaluator/EvaluatorManager.java [hunk 1]
```diff
@@ -487,6 +487,9 @@ public void onResourceLaunch(final ResourceLaunchEvent resourceLaunchEvent) {
       if (this.stateManager.isAllocated()) {
         this.stateManager.setSubmitted();
         this.resourceLaunchHandler.onNext(resourceLaunchEvent);
+      } else if (this.stateManager.isFailedOrKilled()) {
+        LOG.log(Level.WARNING, "Evaluator manager expected" + EvaluatorState.ALLOCATED +
```
Comment (markusweimer) [inline lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/driver/evaluator/EvaluatorManager.java]: Shouldn't this generate a `FailedEvaluator`? Or was that generated prior to reaching this?

Comment (tcNickolas) [inline lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/driver/evaluator/EvaluatorManager.java]: I would expect `FailedEvaluator` to be produced at the code path which set state of this evaluator to failed. This code path doesn't make the decision about failure but merely observes the state set elsewhere.

Comment (markusweimer) [inline lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/driver/evaluator/EvaluatorManager.java]: Ah, yes, that makes sense. The only call to `EvaluatorStatusManager.setFailed` is in `org.apache.reef.runtime.common.driver.evaluator.EvaluatorManager.onEvaluatorException`, which itself is called when we either receive an exception from the Evaluator itself or a shutdown notice from the RM. The latter is the case here.

Comment (markusweimer) [inline lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/driver/evaluator/EvaluatorManager.java]: Back to the code at hand: If this isn't an abnormal code path (it can happen and is handled correctly), shouldn't the log level be `VERBOSE` here?

**Comment by markusweimer** [comment]
Do we want a test for this new behavior? It would be tricky, as we'd have to augment the local runtime to create this situation where an allocated, but not running, Evaluator fails. If we wanted a test, I suggest to file a separate JIRA that links back to this one. The change at hand fixes a bug and should go in before we have a test for this edge case.

File: lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/driver/evaluator/EvaluatorManager.java (hunk 1 above)
Comment (tcNickolas) [inline lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/driver/evaluator/EvaluatorManager.java]: I would keep it as `WARNING`. 1. This occurs infrequently so is not noisy, and 2. If for some reason `FailedEvaluator` event is lost or not processed, this code path will secretly not launch task, and I feel it's better to leave some breadcrumbs for the investigation, rather than hide the decision completely.

**Comment by tcNickolas** [comment]
Agreed, I've filed REEF-1455 for this.

A good AppVeyor build of this code is https://ci.appveyor.com/project/tcNickolas/reef/build/152-REEF-1454

**Comment by jwang98052** [comment]
Chatted with out client, their scenarios is 
Request Evalautor,
call SubmitContextAndServiceAndTask on IAllocatedEvaluator then crash. 

That means the Evalautor is not physically exist yet. Who sets evaluator state to FAILED or KILLED?

**Comment by markusweimer** [comment]
> That means the Evalautor is not physically exist yet. Who sets evaluator state to FAILED or KILLED? 

YARN. It could be that the machine we have the container allocation on crashed before we launched the Evaluator process on it.

File: lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/driver/evaluator/EvaluatorManager.java (hunk 1 above)
Comment (markusweimer) [inline lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/driver/evaluator/EvaluatorManager.java]: Agreed.

**Comment by markusweimer** [comment]
LGTM, will merge

**Comment by jwang98052** [comment]
It is is set from Yarn, will the client still receive FailedEvaluator?

**Comment by markusweimer** [comment]
> It is is set from Yarn, will the client still receive FailedEvaluator? 

I believe so, as I outlined above.
